bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -0,0 +1,392 @@
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.List;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.eclipse.smarthome.core.events.EventPublisher;
+import org.eclipse.smarthome.core.items.events.ItemEventFactory;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.icalendar.internal.config.ICalendarConfiguration;
+import org.openhab.binding.icalendar.internal.handler.PullJob.CalendarUpdateListener;
+import org.openhab.binding.icalendar.internal.logic.AbstractPresentableCalendar;
+import org.openhab.binding.icalendar.internal.logic.CalendarException;
+import org.openhab.binding.icalendar.internal.logic.CommandTag;
+import org.openhab.binding.icalendar.internal.logic.CommandTagType;
+import org.openhab.binding.icalendar.internal.logic.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ICalendarHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Michael Wodniok - Initial contribution
+ *
+ * @author Andrew Fiddian-Green - Support for Command Tags embedded in the Event description
+ *
+ */
+@NonNullByDefault
+public class ICalendarHandler extends BaseThingHandler implements CalendarUpdateListener {
+
+    private final Logger logger = LoggerFactory.getLogger(ICalendarHandler.class);
+    private HttpClient httpClient;
+
+    private @Nullable ICalendarConfiguration configuration;
+    private File calendarFile;
+    private @Nullable AbstractPresentableCalendar runtimeCalendar;
+
+    private @Nullable ScheduledFuture<?> updateJobFuture;
+    private @Nullable ScheduledFuture<?> pullJobFuture;
+
+    @Nullable
+    private EventPublisher eventPublisherCallback = null;
+
+    private Instant updateStatesLastCalledTime = Instant.now();
+
+    public ICalendarHandler(Thing thing, HttpClient httpClient, @Nullable EventPublisher eventPublisher) {
+        super(thing);
+        this.httpClient = httpClient;
+        this.calendarFile = new File(ConfigConstants.getUserDataFolder() + File.separator
+                + getThing().getUID().getAsString().replaceAll("[^a-zA-Z0-9\\._-]", "_") + ".ical");
```

Comment:
The UID can be set freely by user ([docs](https://www.openhab.org/docs/concepts/items.html#state-and-command-type-formatting)). However for the filename i need to make sure to not allow a [bobby tables](https://www.xkcd.com/327/) and thus i'm replacing all characters that are "non-standard" to ensure no path manipulations are possible. Anywhere else this is no problem.

Reply:
I didn't find any references to UIDs in there. But my point is that if a UID can be set by a user, we shouldn't unnecessarily redact their choice. If UIDs support internationalized characters, (e.g. À, à), they should be allowed.

Reply:
ahh this was the wrong buffer, i meant the [configuration guide](https://www.openhab.org/docs/configuration/things.html#defining-things-using-files). To create files i have to ensure no invalid characters are used in the filename, this is what the replacement does, `[a-zA-Z0-9\._-]` should contain only characters which are supported by the most distributed OS'. I could add a part of a cryptographic hash of the UID to minimize the collision probability when a user want's to create a calendar with `à` and `ä`.

Reply:
You shouldn't consider what "the most distributed OS" is.
If you're trying to create a filename-safe string, create a method for it, and remove only the invalid characters.

Reply:
Sure normally i wouldn't consider. But i only know which are the normally accepted characters (i don't know every system by detail but i'm sure enough that all major systems allow the use of uppercase and lowercase "english" letters, numbers, dots, underscores and hyphens). Is there an OpenHAB/System function which i can reuse for normalizing filenames instead so i don't have filter it myself?

Reply:
I know of no built-in method. But I think the best compromise position is to just remove the characters we know are not allowed. This will remove a few more than necessary on *nix, but I think that's ok.

Reply:
I think a blacklist would be much more effort for filtering than a whitelist. You need to filter out ASCII 0x00-0x1f, specific special characters like `\` or `*` and so on, there are many kinds of whitespaces. As the character set is getting greater (e.G. every few months there is a new set of emojis) keeping this up to date may be an endless task and removing only the characters we currently know is error prone.

Please also consider the usage context: This filename is usually not seen by the user and the file never modified. The user may see the filename when he cleans up the user data folder. So if you are unhappy with deriving a human readable filename from the UID, the only solution (without additional configuration) would be a human unreadable derivation like a cryptographic hash.

Reply:
You could use the special \W character in your regex — it is a shortcut for “anything alphanumeric”..

Reply:
Which means A-Z and a-z. Shorter, but not better.

Reply:
Just remove [these illegal](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) characters:

    < (less than)
    > (greater than)
    : (colon)
    " (double quote)
    / (forward slash)
    \ (backslash)
    | (vertical bar or pipe)
    ? (question mark)
    * (asterisk)

Reply:
I've now done that. I'm still not happy with it from technical perspective.